fish-rust/src/env/environment.rs

```diff
@@ -371,8 +379,315 @@
     static ref PRINCIPAL_STACK: EnvStackRef = Arc::new(EnvStack::new());
 }
 
-// Note: this is an incomplete port of env_init(); the rest remains in C++.
-pub fn env_init(do_uvars: bool) {
+/// Some configuration path environment variables.
+const FISH_DATADIR_VAR: &wstr = L!("__fish_data_dir");
+const FISH_SYSCONFDIR_VAR: &wstr = L!("__fish_sysconf_dir");
+const FISH_HELPDIR_VAR: &wstr = L!("__fish_help_dir");
+const FISH_BIN_DIR: &wstr = L!("__fish_bin_dir");
+const FISH_CONFIG_DIR: &wstr = L!("__fish_config_dir");
+const FISH_USER_DATA_DIR: &wstr = L!("__fish_user_data_dir");
+
+/// Maximum length of hostname. Longer hostnames are truncated.
+const HOSTNAME_LEN: usize = 255;
+
+/// Function to get an identifier based on the hostname.
+fn get_hostname_identifier() -> Option<WString> {
+    // The behavior of gethostname if the buffer size is insufficient differs by implementation and
+    // libc version Work around this by using a "guaranteed" sufficient buffer size then truncating
+    // the result.
+    let mut b = [0 as libc::c_char; HOSTNAME_LEN + 1];
+    if unsafe { libc::gethostname(b.as_mut_ptr(), b.len()) } == 0 {
+        let cstr = unsafe { CStr::from_ptr(b.as_ptr()) };
+        let res = str2wcstring(cstr.to_bytes());
+
+        if res.is_empty() {
+            None
+        } else {
+            Some(res)
+        }
+    } else {
+        None
+    }
+}
+
+fn setup_user(vars: &EnvStack) {
+    let uid = unsafe { libc::geteuid() };
+    let user_var = vars.get_unless_empty(L!("USER"));
+
+    let mut userinfo: MaybeUninit<libc::passwd> = MaybeUninit::zeroed();
+    let mut result: *mut libc::passwd = std::ptr::null_mut();
+    let mut buf = [0 as libc::c_char; 8192];
+
+    // If we have a $USER, we try to get the passwd entry for the name.
+    // If that has the same UID that we use, we assume the data is correct.
+    if let Some(user_var) = user_var {
+        let unam_narrow = wcs2zstring(&user_var.as_string());
+        let retval = unsafe {
+            libc::getpwnam_r(
+                unam_narrow.as_ptr(),
+                userinfo.as_mut_ptr(),
+                buf.as_mut_ptr(),
+                buf.len(),
+                &mut result,
+            )
+        };
+        if retval == 0 && !result.is_null() {
+            let userinfo = unsafe { userinfo.assume_init() };
+            if unsafe { *result }.pw_uid == uid {
+                if vars.get_unless_empty(L!("HOME")).is_none() {
+                    if !userinfo.pw_dir.is_null() {
```

Comment:
missing this comment I guess:

	// The uid matches but we still might need to set $HOME.